src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,538 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::token::{Token, TokenProvider};
+use crate::Result;
+use tokio::sync::watch;
+use tokio::time::{sleep, sleep_until, Duration, Instant};
+
+const FIVE_MINUTES: Duration = Duration::from_secs(300);
+const TEN_SECONDS: Duration = Duration::from_secs(10);
+
+#[derive(Debug)]
+pub(crate) struct TokenCache<T>
+where
+    T: TokenProvider + Send + Sync + 'static,
+{
+    rx_token: watch::Receiver<Option<Result<Token>>>,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T: TokenProvider> Clone for TokenCache<T> {
+    fn clone(&self) -> TokenCache<T> {
+        TokenCache {
+            rx_token: self.rx_token.clone(),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+// TODO 1552: Use the token cache in all creds
+#[allow(dead_code)]
+impl<T> TokenCache<T>
+where
+    T: TokenProvider + Send + Sync + 'static,
+{
+    pub(crate) fn new(inner: T) -> Self {
+        let (tx_token, rx_token) = watch::channel::<Option<Result<Token>>>(None);
+
+        tokio::spawn(async move {
+            refresh_task(inner, tx_token).await;
```

Comment:
Observation: this starts fetching tokens at credential construction time. I think that is probably a good feature. The first token could be available without waiting.